include/orb/GroundPropagator.h

```diff
@@ -0,0 +1,173 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+/**
+ * \file GroundPropagator.h
+ * \author Nathan Zimmerberg
+ * \date 23 APR 2020
+ * \brief A class to propagate orbits sent from ground.
+ */
+
+#pragma once
+
+#include "Orbit.h"
+
+namespace orb
+{
+
+/**
+ * Class to propagate orbits sent from ground.
+ * 
+ * The GroundPropagator trys to first minimize the number of grav calls needed to get a 
+ * not propagating orbit estimate, and second use the most recently input Orbit.
+ * 
+ * Implimentation details:
+ * 
+ * Under normal conditions, this estimator just propagates the most recently uplinked Orbit.
+ * in current.
+ * 
+ * If a new Orbit get uplinked it will normally get put in catching_up, and the 
+ * estimator will propagate it to the current time in the background 
+ * while still propagating current as the best estimate. 
+ * Once catching_up is done propagating it replaces current.
+ * 
+ * If another Orbit gets uplinked while catching_up is still being propagated 
+ * in the background, it gets stored in to_catch_up. This ensures too many Orbits getting 
+ * uplinked won't overload the estimator and prevent it from making progress.
+ * If to_catch_up takes fewer grav calls to finish propagating than catching_up 
+ * it replaces catching_up.
+ * Also to_catch_up replaces catching_up if catching_up finishes propagating.
+ */
+class GroundPropagator {
+  private:
+    /** Current Orbit estimate.*/
+    Orbit current;
+
+    /** Orbit catching up to current time.
+     * If catching_up is valid then current is valid.
+     * If catching_up is valid then current.numgravcallsleft()<catching_up.numgravcallsleft().
+     * If catching_up is valid then catching_up is more recently input than current.
+     */
+    Orbit catching_up;
+
+    /** Orbit to catch up to current time once catching_up is done catching up.
+     * If to_catch_up is valid then catching_up is valid.
+     * If to_catch_up is valid then catching_up.numgravcallsleft()<to_catch_up.numgravcallsleft().
+     * If to_catch_up is valid then to_catch_up is more recently input than catching_up.
+     */
+    Orbit to_catch_up;
```

Comment:
While this may be a valid concern I feel like the additional complexity in creating some buffer for orbit objects and then passing around pointers to it may not be worth the extra effort. The relative overhead of a call to `memcpy` compared to the actual propagation is so minimal as well.

Reply:
#183 I made an issue to try this in a future PR if needed.

Reply:
I'm closing the issue because coping will only happen rarely when an Orbit gets caught up or a new Orbit gets uplinked from ground, not every control cycle.